### Ensembled_Technique & Random Forest

There are 3 Ensembled Technique 

1.Bagging

2.Boosting

3.Stacking

#### Bagging ~ Bootstrap Aggregation
It Create Random Sample of Datapoint(Random Train Set) with Sampling with Replacement(A Particular Datapoints can be repeated in multiple Random sample)

Independent of Previous Model (Parallel in nature)

prediction Average / mode is taken

In [1]:
# Bagged Decision Trees for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

#Load the data set
claimants = pd.read_csv("C:/Users/Akaash/Downloads/claimants.csv")

#Dropping the case number columns as it is not required
claimants.drop(["CASENUM"],inplace=True,axis = 1)

# Removing NA values in data set
claimants = claimants.dropna()

# Dividing our data into input and output variables 
X = claimants.iloc[:,1:]
Y = claimants.iloc[:,0]

# Creating K-Fold Instance
seed = 7
kfold = KFold(n_splits=10,random_state=seed,shuffle = True)

#Creating the Instance of DecisionTreeClassifier()
classifier = DecisionTreeClassifier()

#Training the Model -- Bagging
num_trees = 60
#base_estimator= instance of the Technique used which is either DTclassifier or DTRegressor
model = BaggingClassifier(base_estimator=classifier,n_estimators=num_trees,random_state=seed) 

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)


In [2]:
# Accuracy for Each K-Fold
results

array([0.7       , 0.68181818, 0.64545455, 0.64545455, 0.73636364,
       0.70909091, 0.65137615, 0.71559633, 0.65137615, 0.62385321])

In [3]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

67.60383653044204

Inference: Using the Bagging Ensembled Technique the Final Accuracy is 67.60% which is around 4% less from which i got with out Bagging Technique

#### Random Forest
It is Based on the Concept og Bagging where Datapoint are Randomly Sampled.

It Splits Data Point as well as the features Randomly and Creates Decision Tree

In [4]:
# Random Forest Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

#Load the data set
claimants = pd.read_csv("C:/Users/Akaash/Downloads/claimants.csv")

#Dropping the case number columns as it is not required
claimants.drop(["CASENUM"],inplace=True,axis = 1)

# Removing NA values in data set
claimants = claimants.dropna()

# Dividing our data into input and output variables 
X = claimants.iloc[:,1:]
Y = claimants.iloc[:,0]

# Creating K-Fold Instance
kfold = KFold(n_splits=10, random_state=7,shuffle = True)

#Training the Model ~ Random Forest 
num_trees = 100
max_features = 2
model = RandomForestClassifier(n_estimators=num_trees, max_features=max_features) #max_features will sample randomly 2 Features

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)

In [5]:
# Accuracy for Each K-Fold
results

array([0.73636364, 0.7       , 0.60909091, 0.64545455, 0.70909091,
       0.72727273, 0.69724771, 0.77981651, 0.64220183, 0.63302752])

In [6]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

68.79566305254379

Inference: Using The Random Forest the Final Accuracy is 69.43%

#### Boosting ~ AdaBoost
Stage Weight of That Datapoint Who give Error is updated in the next Training Set of Next model, and the Next Model Try to minimize the Error / Boost the Performance of the model

It is Dependent on the Previous model i.e sequential in Nature

In [7]:
# AdaBoost Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import AdaBoostClassifier

#Load the data set
claimants = pd.read_csv("C:/Users/Akaash/Downloads/claimants.csv")

#Dropping the case number columns as it is not required
claimants.drop(["CASENUM"],inplace=True,axis = 1)

# Removing NA values in data set
claimants = claimants.dropna()

# Dividing our data into input and output variables 
X = claimants.iloc[:,1:]
Y = claimants.iloc[:,0]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

#Training the Model ~ Adaboost (Boosting)
num_trees = 15
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)


In [8]:
# Accuracy for Each K-Fold
results

array([0.81818182, 0.70909091, 0.68181818, 0.71818182, 0.77272727,
       0.70909091, 0.74311927, 0.73394495, 0.71559633, 0.67889908])

In [9]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

72.80650542118433

Inference: Using the Boosting Ensembled Technique the Final Accuracy is 72.80%

#### Stacking

Through Stacking Can Use Multiple Algorithms At Once the TAle mean/Mode of the Prediction

In [10]:
# Stacking Ensemble for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier

#Load the data set
claimants = pd.read_csv("C:/Users/Akaash/Downloads/claimants.csv")

#Dropping the case number columns as it is not required
claimants.drop(["CASENUM"],inplace=True,axis = 1)

# Removing NA values in data set
claimants = claimants.dropna()

# Dividing our data into input and output variables 
X = claimants.iloc[:,1:]
Y = claimants.iloc[:,0]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

# create the sub models - 3 Different Models
estimators = []
model1 = LogisticRegression(max_iter=500)
estimators.append(('logistic', model1))
model2 = DecisionTreeClassifier()
estimators.append(('cart', model2))
model3 = SVC()
estimators.append(('svm', model3))

# create the ensemble model
ensemble = VotingClassifier(estimators)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(ensemble, X, Y, cv=kfold)


In [11]:
# Accuracy for Each K-Fold
results

array([0.72727273, 0.70909091, 0.60909091, 0.72727273, 0.70909091,
       0.7       , 0.67889908, 0.68807339, 0.71559633, 0.71559633])

In [12]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

69.79983319432861

Inference: Using the Stacking Ensembled Technique the Final Accuracy is 69.61%